src/measure_theory/covering/liminf_limsup.lean

```diff
@@ -0,0 +1,196 @@
+  replace hMr : ∀ᶠ j in at_top, M * r₁ (f j) ≤ r₂ (f j) := hf₃.eventually hMr,
+  replace hf₀ : ∀ j, ∃ (w ∈ s (f j)), d ∈ closed_ball w (r₁ (f j)) := λ j,
+    by simpa only [(hs (f j)).cthickening_eq_bUnion_closed_ball (hrp (f j)), mem_Union] using hf₀ j,
+  choose w hw hw' using hf₀,
+  let C := is_doubling_measure.scaling_constant_of μ M⁻¹,
+  have hC : 0 < C :=
+    lt_of_lt_of_le zero_lt_one (is_doubling_measure.one_le_scaling_constant_of μ M⁻¹),
+  suffices : ∃ (η < (1 : ℝ≥0)), ∀ᶠ j in at_top,
+    μ (W ∩ closed_ball (w j) (r₁ (f j))) / μ (closed_ball (w j) (r₁ (f j))) ≤ η,
+  { obtain ⟨η, hη, hη'⟩ := this,
+    replace hη' : 1 ≤ η := by simpa only [ennreal.one_le_coe_iff] using
+      le_of_tendsto (hd' w (λ j, r₁ (f j)) hr $ eventually_of_forall (by simpa only [one_mul])) hη',
+    exact (lt_self_iff_false _).mp (lt_of_lt_of_le hη hη'), },
+  refine ⟨1 - C⁻¹, tsub_lt_self zero_lt_one (nnreal.inv_pos.mpr hC), _⟩,
+  replace hC : C ≠ 0 := ne_of_gt hC,
+  let b : ℕ → set α := λ j, closed_ball (w j) (M * r₁ (f j)),
+  let B : ℕ → set α := λ j, closed_ball (w j) (r₁ (f j)),
+  have h₁ : ∀ j, b j ⊆ B j :=
+    λ j, closed_ball_subset_closed_ball (mul_le_of_le_one_left (hrp (f j)) hM'.le),
+  have h₂ : ∀ j, W ∩ B j ⊆ B j := λ j, inter_subset_right W (B j),
+  have h₃ : ∀ᶠ j in at_top, disjoint (b j) (W ∩ B j),
+  { apply hMr.mp,
+    rw eventually_at_top,
+    refine ⟨i, λ j hj hj', disjoint.inf_right (B j) $ disjoint.inf_right' (blimsup Y₁ at_top p) _⟩,
+    change disjoint (b j) (Z i)ᶜ,
+    rw disjoint_compl_right_iff_subset,
+    refine (closed_ball_subset_cthickening (hw j) (M * r₁ (f j))).trans
+      ((cthickening_mono hj' _).trans (λ a ha, _)),
+    simp only [mem_Union, exists_prop],
+    exact ⟨f j, ⟨hf₁ j, hj.le.trans (hf₂ j)⟩, ha⟩, },
+  have h₄ : ∀ᶠ j in at_top, μ (B j) ≤ C * μ (b j) :=
+    (hr.eventually (is_doubling_measure.eventually_measure_le_scaling_constant_mul'
+      μ M hM)).mono (λ j hj, hj (w j)),
+  refine (h₃.and h₄).mono (λ j hj₀, ennreal.div_le_of_le_mul _),
+  change μ (W ∩ B j) ≤ ↑(1 - C⁻¹) * μ (B j),
+  have hB : μ (B j) ≠ ∞ := measure_closed_ball_lt_top.ne,
+  rw [with_top.coe_sub, ennreal.coe_one, ennreal.sub_mul (λ _ _, hB), one_mul],
+  replace hB : ↑C⁻¹ * μ (B j) ≠ ∞,
+  { refine ennreal.mul_ne_top _ hB,
+    rwa [ennreal.coe_inv hC, ne.def, ennreal.inv_eq_top, ennreal.coe_eq_zero], },
+  obtain ⟨hj₁ : disjoint (b j) (W ∩ B j), hj₂ : μ (B j) ≤ C * μ (b j)⟩ := hj₀,
+  replace hj₂ : ↑C⁻¹ * μ (B j) ≤ μ (b j),
+  { rw [ennreal.coe_inv hC, ← ennreal.div_eq_inv_mul],
+    exact ennreal.div_le_of_le_mul' hj₂, },
+  have hj₃ : ↑C⁻¹ * μ (B j) + μ (W ∩ B j) ≤ μ (B j),
+  { refine le_trans (add_le_add_right hj₂ _) _,
+    rw ← measure_union' hj₁ measurable_set_closed_ball,
+    exact measure_mono (union_subset (h₁ j) (h₂ j)), },
+  replace hj₃ := tsub_le_tsub_right hj₃ (↑C⁻¹ * μ (B j)),
+  rwa ennreal.add_sub_cancel_left hB at hj₃,
+end
+
+/-- This lemma is a generalisation of Lemma 9 appearing on page 217 of
```

Comment:
I'd be happy to have a docstring explaining what the lemma is saying, because this is not a trivial result: there is something interesting going on here, so it's worth pointing it out. And you can even call it a theorem :-)

Reply:
Done :-)